core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DocumentVersionInfo.java

```diff
@@ -19,57 +19,181 @@
 import static com.mongodb.stitch.core.services.mongodb.remote.sync.internal.DataSynchronizer.DOCUMENT_VERSION_FIELD;
 
 import com.mongodb.stitch.core.internal.common.BsonUtils;
+
+import java.util.UUID;
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import org.bson.BsonBoolean;
 import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonInt64;
+import org.bson.BsonString;
 import org.bson.BsonValue;
 
+
 final class DocumentVersionInfo {
-  private final BsonValue version;
-  private final BsonDocument filter;
+  @Nullable private final VersionValues versionValues;
+  @Nullable private final BsonDocument versionDoc;
+  @Nullable private final BsonDocument filter;
+
+  private class VersionValues {
+    final int syncProtocolVersion;
+    final String instanceId;
+    final long versionCounter;
+
+    VersionValues(
+            final int syncProtocolVersion,
+            final String instanceId,
+            final long versionCounter) {
+      this.syncProtocolVersion = syncProtocolVersion;
+      this.instanceId = instanceId;
+      this.versionCounter = versionCounter;
+    }
+  }
+
+  private static final String SYNC_PROTOCOL_VERSION_FIELD = "spv";
+  private static final String INSTANCE_ID_FIELD = "id";
+  private static final String VERSION_COUNTER_FIELD = "v";
+
+  private DocumentVersionInfo(
+          @Nullable final BsonDocument version,
+          @Nullable final BsonValue documentId
+  ) {
+    if (version != null) {
+      this.versionDoc = version;
+      this.versionValues = new VersionValues(
+        versionDoc.getInt32(SYNC_PROTOCOL_VERSION_FIELD).getValue(),
+        versionDoc.getString(INSTANCE_ID_FIELD).getValue(),
+        versionDoc.getInt64(VERSION_COUNTER_FIELD).getValue()
+      );
+    } else {
+      this.versionDoc = null;
+      this.versionValues = null;
+    }
+
+    if (documentId != null) {
+      this.filter = getVersionedFilter(documentId, version);
+    } else {
+      this.filter = null;
+    }
+  }
+
+  @Nullable BsonDocument getVersionDoc() {
+    return versionDoc;
+  }
+
+  /**
+   * Returns whether this version is non-empty (i.e. a version from a document with no version)
+   * @return true if the version is non-empty, false if the version is empty.
+   */
+  boolean isNonEmptyVersion() {
+    return versionValues != null;
+  }
+
+  private @Nonnull VersionValues getVersionValuesThrowIfNull() {
```

Comment:
[minor] Using `ThrowIfNull` as a suffix here, and the exception thrown here, is arguably a bit of an anti-pattern. If this _is_ an illegal state, we should throw that exception in the constructor– otherwise, it's on the user to call the `#isNonEmptyVersion` check.